sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java

```diff
@@ -96,13 +98,35 @@ public UnsafeKVExternalSorter(
         numElementsForSpillThreshold,
         canUseRadixSort);
     } else {
-      // The array will be used to do in-place sort, which require half of the space to be empty.
-      assert(map.numKeys() <= map.getArray().size() / 2);
+      // Becasue we insert the number of values in the map into `UnsafeInMemorySorter`, if
+      // the number of values is more than the number of keys, and the array in the map is
+      // not big enough to do in-place sort, we must acquire new array.
+      // To insert a record into `UnsafeInMemorySorter` will consume two spaces in the array.
+      // We must have half of the array as empty. There are totally `map.numValues()` records
+      // to be inserted.
+      LongArray sortArray = null;
+      boolean useAllocatedArray = false;
+      if (map.numValues() > map.numKeys() && map.numValues() * 2 > map.getArray().size() / 2) {
```

Comment:
If you ask the reason behind this calculation, it is because the a value needs two spaces (pointer, key's hashcode) in the array. We need at least half of the array for sorting. So if the occupied space of all values are more than half of the array, we need to allocate new bigger array for sorting.